### PR TITLE
Validate build hash in handshake

### LINKED ISSUE #65249: Validate build hash of joining node
Today our more intrepid users sometimes form a cluster from unreleased versions in order to test out new functionality. This is usually fine, and indeed is encouraged, but can lead to difficulties if they happen to use two different builds of the same numbered version since there is no guarantee of compatibility between such builds. Clusters like this can behave quite strangely and it can be tricky to work out why. A similar issue occurs when using a remote cluster with an equal version but a different build hash.

I'm opening this issue to discuss whether we should prevent an invalid mix of nodes from forming a faulty cluster or remote cluster connection by sharing `Build.CURRENT.hash()` as well as `Version.CURRENT` somewhere in the handshaking process. If two nodes determine that they have the same numbered version, but a different build hash, then they would emit a warning and drop the connection.

### PR DESCRIPTION
**Pull request by DaveCTurner**
There is no guarantee of wire compatibility between nodes running
different builds of the same version, but today we do not validate
whether two communicating nodes are compatible or not. This results in
confusing failures that look like serialization bugs, and it usually
takes nontrivial effort to determine that the failure is in fact due to
the user running incompatible builds.

This commit adds the build hash to the transport service handshake and
validates that matching versions have matching build hashes.

Closes #65249

### CODE REVIEW

**Comment by DaveCTurner** [comment]
NB I opened this PR against `7.x` because it's slightly more involved there thanks to the transport client.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 1]
```diff
@@ -73,10 +75,14 @@
...
 
-public class TransportService extends AbstractLifecycleComponent implements ReportingService<TransportInfo>, TransportMessageListener,
-    TransportConnectionListener {
+public class TransportService extends AbstractLifecycleComponent
+        implements ReportingService<TransportInfo>, TransportMessageListener, TransportConnectionListener {
+
     private static final Logger logger = LogManager.getLogger(TransportService.class);
 
+    private static final String PERMIT_HANDSHAKES_FROM_INCOMPATIBLE_BUILDS_KEY = "es.unsafely_permit_handshake_from_incompatible_builds";
+    private static final boolean PERMIT_HANDSHAKES_FROM_INCOMPATIBLE_BUILDS = getPermitHandshakesFromIncompatibleBuilds();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:84]: NIT: maybe just put the logic in this method in a `static { }` block below these two constants to make this easier to read in one go?

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 2]
```diff
@@ -192,7 +205,14 @@ public TransportService(Settings settings, Transport transport, ThreadPool threa
             false, false,
             HandshakeRequest::new,
             (request, channel, task) -> channel.sendResponse(
-                new HandshakeResponse(localNode, clusterName, localNode.getVersion())));
+                new HandshakeResponse(localNode.getVersion(), Build.CURRENT.hash(), localNode, clusterName)));
+
+        if (PERMIT_HANDSHAKES_FROM_INCOMPATIBLE_BUILDS) {
+            logger.warn("transport handshakes from incompatible builds are unsafely permitted on this node; remove system property [" +
+                    PERMIT_HANDSHAKES_FROM_INCOMPATIBLE_BUILDS_KEY + "] to resolve this warning");
+            DeprecationLogger.getLogger(TransportService.class).deprecate("permit_handshake_from_incompatible_builds",
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:213]: NIT: little much to log two messages for one thing? This seems like it's kind of a dev-only option anyway, why bother with the deprecation logger?

**Review by original-brownbear — APPROVED** [review]
LGTM, just random NITs that can be ignored :)

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:226]: Ehh I worry we'll get people using this, and the two logs are for two separate aspects. Deprecation logs don't go into the main server log (by default at least) but I think this deserves a visible warning whenever it's used.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java]: Ah yeah not sure why I didn't do that. Done in 8132bf33b0b889e84d601a18421fd894e0ed8e36

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Pull request merged**